web/src/engine/predictive-text/worker-thread/src/main/predict-helpers.ts

```diff
@@ -58,24 +58,77 @@ export const CORRECTION_SEARCH_THRESHOLDS = {
   REPLACEMENT_SEARCH_THRESHOLD: 4 as const // e^-4 = 0.0183156388.  Allows "80%" of an extra edit.
 }
 
+/**
+ * Collates information related to suggestions during the suggestion generation
+ * process.
+ */
 export type CorrectionPredictionTuple = {
-  prediction: ProbabilityMass<Suggestion>,
+  /**
+   * The potential Suggestion (or Keep)
+   */
+  prediction: ProbabilityMass<Suggestion | Keep>,
+  /**
+   * The correction upon which the Suggestion (or Keep) is based
+   */
   correction: ProbabilityMass<string>,
+  /**
+   * The likelihood of the prediction - its lexical-model likelihood multiplied
+   * by the keystroke-sequence + correction likelihood.
+   */
   totalProb: number;
-  matchLevel: SuggestionSimilarity;
+  /**
+   * How directly the prediction matches the current token in the context.
+   *
+   * This is determined later in the suggestion-analysis project and is not
+   * available upon initial construction of this type.
+   */
+  matchLevel?: SuggestionSimilarity;
+  /**
+   * Text from the triggering input that should _not_ be affected by the
+   * prediction.
+   */
   preservationTransform?: Transform;
 };
 
+/**
+ * An enum to be used when categorizing the level of similarity between
+ * generated Suggestions and the actual text upon which a Suggestion is
+ * based.
+ *
+ * Note that Suggestions require .exact matching to stand-in as the Keep
+ * option.
+ */
 export enum SuggestionSimilarity {
+  /**
+   * The keyed form for the current token / word does not match
+   * the keyed form of the suggestion.
+   */
   none = 0,
+
+  /**
+   * The keyed form for the current token / word matches the
+   * the keyed form of the suggestion, but they do not match
+   * in a case-insensitive manner.
+   */
   sameKey = 1,
+
+  /**
+   * The current token / word has a case-insensitive match with
+   * the suggestion, but not a case-sensitive match.  Both share
+   * the same keyed form.
+   */
   sameText = 2,
+
+  /**
+   * The current token / word has a case-sensitive match with
+   * the suggestion in addition to sharing the same keyed form.
+   */
   exact = 3
 }
 
 export function tupleDisplayOrderSort(a: CorrectionPredictionTuple, b: CorrectionPredictionTuple) {
   // Similarity distance
-  const simDist = b.matchLevel - a.matchLevel;
+  const simDist = b.matchLevel ?? 0 - a.matchLevel ?? 0;
```

Comment:
operator precedence!

```suggestion
  const simDist = (b.matchLevel ?? 0) - (a.matchLevel ?? 0);
```

```
> undefined ?? 0 - undefined ?? 0
Nan
> (undefined ?? 0) - (undefined ?? 0)
0
```